### PR TITLE
Options: support array value paths for panel options

### PR DESCRIPTION
**Pull request by ryantxu**
This is a small part of #39491 -- and may actually support many of the use-cases that require complex objects.  This lets you set an array style path and it will be managed properly:

image

will create:

image

### CODE REVIEW

File: public/app/features/dashboard/components/PanelEditor/utils.ts [hunk 1]
```diff
@@ -68,6 +68,15 @@ export function setOptionImmutably<T extends object>(options: T, path: string |
...
+    let idx = key.lastIndexOf('[');
+    const index = +key.substring(idx + 1, key.length - 1);
+    const propKey = key.substr(0, idx);
+    const current = (options as Record<string, any>)[propKey];
+    const arr = Array.isArray(current) ? [...current] : []; // note spread behavior for missing values is preferable
+    arr[index] = splat.length ? setOptionImmutably(current, splat, value) : value;
+    return { ...options, [propKey]: arr };
+  }
 
   if (!splat.length) {
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/PanelEditor/utils.ts:81]: I guess this is the break condition for the recursiveness, can it be placed higher up in the function?

**Review by hugohaggmark — APPROVED** [review]
The code looks, I did a high level code review as I don't fully understand how this is going to be used.

**Review by torkelo — COMMENTED** [review]
What are the use cases for options defined at a specific array index? could not an object with prop names be used to solve the same problem:

```ts
sample: {
 "a1": {},
 "a2": {}
}
```


And I think we use mergeWith to merge defaults and if there is an array prop we take the whole array so no new defaults will be applied. 

This make me a little bit worried that it complicates the already complex option models. Will require an update to the clean props & apply defaults code. 
https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/state/getPanelOptionsWithDefaults.ts#L113

**Comment by ryantxu** [comment]
> What are the use cases for options defined at a specific array index? 

Well supporting arrays in general.  Arrays as options are good for things that represent an ordered list ;)  

This PR is a small part of #39491 -- that will immediately let us work with things of list size 1 without the need for a whole custom editor.  In #39491, the the panel options builder is called with context and can respond to changes.

Current places where list size 0 would be nice to avoid custom editor while we sort out nested options lists are:
* geomap currently supports single layer only (need to make it multi soon)
* styles within geomap layer
* scatter plot 'series' config

**Comment by ryantxu** [comment]
@torkelo -- I added an error if you try to specify a fieldConfig with array.  That will let us avoid the prop cleaning issues for field configs.

With panel options we clear them all anyway so I think it is ok

**Comment by torkelo** [comment]
> Well supporting arrays in general. Arrays as options are good for things that represent an ordered list ;)

But if the order is defined in advance in the code why does it matter if it's using an array or numbered props?

**Comment by ryantxu** [comment]
Ahh I get your question -- yes right now this is useful only for index "[0]" and avoiding needing to define a new migration when we more easily support nested options #39491.

File: public/app/features/dashboard/components/PanelEditor/utils.ts (hunk 1 above)
Comment (ryantxu) [reply public/app/features/dashboard/components/PanelEditor/utils.ts:81]: note that the array case above also has a return.  (with the added behavior of forcing array behavior)

**Review by dprokop — APPROVED** [review]
Think this looks good @ryantxu. I've shared similar concerns to the ones that @torkelo mentioned, but I guess I get the point of array syntax now when you explained it. Do you intend to use this change for indexes other than 0?

**Review by torkelo — APPROVED** [review]
Ok, had a chat with Ryan yesterday to get a better understanding of how this is to be used and why it would be useful.

**Pull request merged**